### PR TITLE
feat: revoke approvals

### PR DESCRIPTION
**Pull request by estebanmino**
Fixes BX-1306
Figma link (if any):

https://www.figma.com/file/hzfwKOddfRmeVO4KUsG9wM/branch/EP18mAGiXp83ruTOrai9wJ/Rainbow-BX?node-id=15843%3A386013&mode=dev

## What changed (plus any additional context for devs)

implemented approvals flows following figma

https://www.loom.com/share/eebc3bf57f8d4cdbb93ac1969bce9569

### CODE REVIEW

**Review by BrodyHughes — CHANGES_REQUESTED** [review]
The flow in general looks / works great. Have a few considerations that can be tracked here: https://linear.app/rainbow/issue/BX-1306/approvals-qa

File: src/entries/popup/pages/settings/settings.tsx [hunk 1]
```diff
@@ -199,43 +201,63 @@ export function Settings() {
...
+            onClick={() => navigate(ROUTES.SETTINGS__TRANSACTIONS)}
             titleComponent={
-              <MenuItem.Title text={i18n.t('settings.networks.title')} />
+              <MenuItem.Title text={i18n.t('settings.transactions.title')} />
             }
+            testId="settings-transactions"
           />
+          {isWatchingWallet ? null : (
+            <MenuItem
+              first
```
Comment (BrodyHughes) [inline src/entries/popup/pages/settings/settings.tsx:223]: I think this need to have `last` here instead. I realized the tab-ing works, bit the highlight border radius are slightly wrong

**Review by BrodyHughes — APPROVED** [review]
with those other changes it looks good.

File: src/entries/popup/pages/settings/settings.tsx (hunk 1 above)
Comment (estebanmino) [reply src/entries/popup/pages/settings/settings.tsx]: yeah i just updated thx

File: src/core/resources/gas/estimateApprovalGasLimit.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+// ///////////////////////////////////////////////
+// Query Types
+
+export type EstimateAprovalGasLimitArgs = {
+  chainId: ChainId;
+  ownerAddress: Address;
+  assetAddress?: Address;
+  spenderAddress?: Address;
+  assetType: 'erc20' | 'erc721';
```
Comment (greg-schrammel) [inline src/core/resources/gas/estimateApprovalGasLimit.ts:24]: no support for erc1155?

File: src/entries/popup/pages/home/Activity/ActivitiesList.tsx [hunk 3]
```diff
@@ -48,9 +58,71 @@ export function Activities() {
...
+  const { currentAddress } = useCurrentAddressStore();
+  const { currentCurrency } = useCurrentCurrencyStore();
+  const { rainbowChains } = useRainbowChains();
+  const supportedMainnetIds = SUPPORTED_MAINNET_CHAINS.map((c: Chain) => c.id);
+  const { userChains } = useUserChainsStore();
+  const { isWatchingWallet } = useWallets();
+
+  const chainIds = rainbowChains
+    .filter((c) => supportedMainnetIds.includes(c.id) && userChains[c.id])
+    .map((c) => c.id);
```
Comment (greg-schrammel) [inline src/entries/popup/pages/home/Activity/ActivitiesList.tsx:70]: do we need the rainbowChains? can't we filter filter the supportedMainnetIds from userChains?

File: src/entries/popup/pages/home/Activity/ActivitiesList.tsx [hunk 4]
```diff
@@ -48,9 +58,71 @@ export function Activities() {
...
+    (tx: RainbowTransaction) => {
+      if (tx.type === 'approve' && !isWatchingWallet) {
+        const txApproval = tokenApprovals?.find((approval) =>
+          isLowerCaseMatch(approval.spender.tx_hash, tx.hash),
+        );
+        if (txApproval) {
+          return true;
+        }
+      }
+      return false;
```
Comment (greg-schrammel) [inline src/entries/popup/pages/home/Activity/ActivitiesList.tsx:103]: it's just my opinion but I would find something like this easier to read
```suggestion
      if (tx.type !== 'approve' || !isWatchingWallet) return false
      return tokenApprovals?.some((approval) =>
        isLowerCaseMatch(approval.spender.tx_hash, tx.hash),
      );
```

File: src/entries/popup/pages/home/Activity/ActivityContextMenu.tsx [hunk 5]
```diff
@@ -50,27 +52,41 @@ export function ActivityContextMenu({
...
+    revokeRef?.current?.click();
+    onRevokeTransaction?.();
+  }, [onRevokeTransaction]);
+
+  useKeyboardShortcut({
+    condition: () => !!open,
+    handler: (e: KeyboardEvent) => {
+      e.preventDefault();
+      if (e.key === shortcuts.activity.REFRESH_TRANSACTIONS.key) {
+        // if (onRevokeTransaction) onRevokeTransaction();
```
Comment (greg-schrammel) [inline src/entries/popup/pages/home/Activity/ActivityContextMenu.tsx:76]: comment

File: src/entries/popup/pages/home/Approvals/utils.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  callback?: () => void;
+}
+
+export const listenRevokeApproval = (
+  callback: ({ callback, show, approval }: RevokeApprovalProps) => void,
+) => {
+  eventEmitter.addListener('rainbow_revoke_approval', callback);
+  return () => {
+    eventEmitter.removeListener(
+      'rainbow_app_connection_wallet_switcher',
```
Comment (greg-schrammel) [inline src/entries/popup/pages/home/Approvals/utils.tsx:25]:
```suggestion
      'rainbow_revoke_approval',
```

**Review by greg-schrammel — APPROVED** [review]
looks good

**Review by derHowie — APPROVED** [review]
👌 🔥

File: src/entries/popup/pages/home/Activity/ActivitiesList.tsx (hunk 3 above)
Comment (estebanmino) [reply src/entries/popup/pages/home/Activity/ActivitiesList.tsx:70]: hm yeah i think that would work

File: src/core/resources/gas/estimateApprovalGasLimit.ts (hunk 2 above)
Comment (estebanmino) [reply src/core/resources/gas/estimateApprovalGasLimit.ts:24]: this is just a local param 

https://github.com/rainbow-me/browser-extension/pull/1322/files#diff-ecd89036959c55e3706a5919f309eba47e8e25aaef429bdda06fcb2cc83eefebR101

backend is returning 'nft' so i'm gonna do the same to have it cleaner

**Pull request merged**